src/gt4py/next/iterator/transforms/unroll_reduce.py

```diff
@@ -133,11 +140,7 @@ class UnrollReduce(NodeTranslator):
     def apply(cls, node: itir.Node, **kwargs):
         return cls().visit(node, **kwargs)
 
-    def visit_FunCall(self, node: itir.FunCall, **kwargs):
-        node = self.generic_visit(node, **kwargs)
-        if not _is_reduce(node):
-            return node
-
+    def _visit_reduce(self, node: itir.FunCall, **kwargs) -> itir.Expr:
```

Comment:
The suggestion is correct, but what I care is that it is `itir.Expr`. Should I still maximally constrain? I want that users of this function assume it's an Expr not a specific Expr (e.g. we could refactor such that it return `init` if `max_neighbors == 0`).

Reply:
My question would be: is it Python best practices to use always the most derived type or instead document what the user of my function should assume about the return type?
The refactoring argument would be, with `FunCall`, users of my function can assume more than with `Expr`, so I might have to change not only my function but would have to check on all call sites. Again, not super relevant for this case, but I would like to understand what the best practices are in general.

Reply:
To get as safe as possible type checks, the best practice is to be as specific as possible in the annotations of the outputs and as permissive as possible in the inputs of a function. The current implementation always returns `itir.FunCall`s but you say that this is just an implementation detail and the (implicit) semantics are that the caller should treat the returned value as an `itir.Expr`, right? In this case I think you're right and the most appropriate annotation would be `Expr`, because we want the type checker to catch any use of the returned value where an `Expr` is not good enough.

Reply:
Thanks!